Adding disk buffering, part 1

LikeTheSalad opened the pull request:
Related to #132 

Setting up the configuration and utilities needed for adding the disk buffering/cache support in a follow-up PR.

LikeTheSalad commented on instrumentation/src/main/java/io/opentelemetry/android/internal/services/Service.java:
This is related to #129 

At the moment is pretty similar to what we have at Elastic, though I'm open to suggestions/alternatives.

breedx-splk commented on instrumentation/src/main/java/io/opentelemetry/android/internal/services/AppInfoService.java:
```diff
@@ -0,0 +1,93 @@
+ * <p>This class is internal and not for public use. Its APIs are unstable and can change at any
+ * time.
+ */
+public class AppInfoService implements Service {
```
Looking at the public surface/outline of this class, it doesn't _really_ seem to deal with `AppInfo` very much....it deals more with the cache. Can we call it `CacheStorageService` instead?

LikeTheSalad replied:
Sure, makes sense. It's updated now.

breedx-splk commented on instrumentation/src/main/java/io/opentelemetry/android/internal/services/AppInfoService.java:
```diff
@@ -0,0 +1,93 @@
+    }
+
+    @WorkerThread
+    public long getAvailableCacheSpace(long maxSpaceNeeded) {
```
It can be frustrating to troubleshoot a problem and then later discover that a seemingly harmless "getter" method has side effects (like allocating cache memory here). I think a rename like `ensureCacheSpaceAvailable(long)` helps with this -- that way the caller is more aware. 

A javadoc is also warranted. Right now the reader has no clue what the return value means.

LikeTheSalad replied:
Thanks, it's updated now.

breedx-splk commented on instrumentation/src/main/java/io/opentelemetry/android/internal/services/ServiceManager.java:
```diff
@@ -0,0 +1,77 @@
+
+public final class ServiceManager implements Lifecycle {
+
+    private final HashMap<Service.Type, Service> services = new HashMap<>();
```
```suggestion
    private final Map<Service.Type, Service> services = new HashMap<>();
```

Is concurrency important to this class?

LikeTheSalad replied:
> Is concurrency important to this class?

Not for now, ideally the writing to this map should happen once, during the rum initialization process.

breedx-splk commented on instrumentation/src/main/java/io/opentelemetry/android/internal/services/Service.java:
```diff
@@ -0,0 +1,16 @@
+
+package io.opentelemetry.android.internal.services;
+
+public interface Service extends Lifecycle {
```
Can we default the `start()` and `stop()` methods to no-op?

LikeTheSalad replied:
Sure, sometimes I forget we can do that... It's updated now.

breedx-splk commented on instrumentation/src/main/java/io/opentelemetry/android/internal/services/PreferencesService.java:
```diff
@@ -0,0 +1,49 @@
+public class PreferencesService implements Service {
+    private final SharedPreferences preferences;
+
+    public PreferencesService(Context context) {
+        preferences =
+                context.getSharedPreferences(
+                        BuildConfig.LIBRARY_PACKAGE_NAME + ".prefs", Context.MODE_PRIVATE);
+    }
```
Code smell -- wrong dependency passed to constructor. I'd make this a factory method `create(context)` and pass the `SharedPreferences` instance into a new constructor.

marandaneto replied:
Ideally this is also lazily loaded (or off of the main thread) instead of in the ctor since this does IO and will be warned by `StrictMode`.

LikeTheSalad replied:
> I'm assuming this will be created once the SDK starts on the main thread

You're right. The idea is to initialize it in `OpenTelemetryRumBuilder` which should be called in the App.onCreate method. We're going to have to take a look at what parts of the initialization need to happen in a worker thread.

breedx-splk replied:
Good call @marandaneto. Would you mind opening an issue to track this so that we don't have to force it into this sizable PR? Thanks.

breedx-splk replied:
Ok, I've went ahead and logged #149 for this.

breedx-splk commented on instrumentation/src/main/java/io/opentelemetry/android/internal/features/persistence/DiskManager.java:
```diff
@@ -0,0 +1,104 @@
+
+    public int getMaxFolderSize() {
+        int storedSize = preferencesService.retrieveInt(MAX_FOLDER_SIZE_KEY, -1);
+        if (storedSize != -1) {
```
nit: if somehow another negative number ended up in there we would return it. How about `if (storedSize > 0)`?

LikeTheSalad replied:
Good point! it's updated now.

breedx-splk commented on instrumentation/src/main/java/io/opentelemetry/android/internal/features/persistence/DiskManager.java:
```diff
@@ -0,0 +1,104 @@
+        }
+        int requestedSize = persistenceConfiguration.maxCacheSize;
+        int availableCacheSize = (int) appInfoService.getAvailableCacheSpace(requestedSize);
+        int calculatedSize = (availableCacheSize / 3) - MAX_FILE_SIZE;
```
I think we should guard against this going negative. Maybe it needs to be ``>= 2*MAX_FILE_SIZE` as well?

LikeTheSalad replied:
You're right, I've added some javadoc and a validation for it. It might need some polish in the future though.

breedx-splk reviewed (commented):
This looks like a good start, thanks!

marandaneto commented on instrumentation/src/main/java/io/opentelemetry/android/internal/features/persistence/DiskManager.java:
```diff
@@ -0,0 +1,104 @@
+
+    public File getSignalsCacheDir() throws IOException {
+        File dir = new File(appInfoService.getCacheDir(), "opentelemetry/signals");
+        if (!dir.exists()) {
```
`mkdirs` already calls `exists` internally so you are doing IO twice.
To consider is that `mkdirs` returns `false` if `exists` returns true, so you cannot `throw IOException` in this case which is also suboptinal.

marandaneto replied:
`Files.deleteIfExists` does that better but its probably not available under older Android versions.

LikeTheSalad replied:
> `Files.deleteIfExists` does that better but its probably not available under older Android versions.

Yeah, it seems like it's available from API 26 😅

LikeTheSalad commented on gradle/libs.versions.toml:
```diff
@@ -25,7 +25,8 @@ opentelemetry-api = { module = "io.opentelemetry:opentelemetry-api" }
 opentelemetry-exporter-logging = { module = "io.opentelemetry:opentelemetry-exporter-logging" }
-zipkin-sender-okhttp3 = { module = "io.zipkin.reporter2:zipkin-sender-okhttp3", version.ref = "zipkin-reporter"}
+zipkin-sender-okhttp3 = { module = "io.zipkin.reporter2:zipkin-sender-okhttp3", version.ref = "zipkin-reporter" }
+opentelemetry-diskBuffering = "io.opentelemetry.contrib:opentelemetry-disk-buffering:1.31.0-alpha"
```
Thanks, it's updated now.

LikeTheSalad commented on instrumentation/src/main/java/io/opentelemetry/android/config/OtelRumConfig.java:
```diff
@@ -90,4 +93,16 @@ public OtelRumConfig disableScreenAttributes() {
+
+    public PersistenceConfiguration getPersistenceConfiguration() {
+        if (persistenceConfiguration == null) {
+            return PersistenceConfiguration.builder().build();
```
jsyk, this was deleted after making the config field not nullable.

LikeTheSalad commented on instrumentation/src/main/java/io/opentelemetry/android/config/OtelRumConfig.java:
```diff
@@ -20,6 +22,7 @@ public class OtelRumConfig {
     private boolean includeNetworkAttributes = true;
     private boolean generateSdkInitializationEvents = true;
     private boolean includeScreenAttributes = true;
+    @Nullable private PersistenceConfiguration persistenceConfiguration = null;
```
Sounds good, I've made it not nullable and set it to a new, default instance in order to reduce memory allocation.

LikeTheSalad commented on instrumentation/src/main/java/io/opentelemetry/android/config/PersistenceConfiguration.java:
```diff
@@ -0,0 +1,51 @@
+    }
+
+    public static Builder builder() {
+        return new Builder(false, 60 * 1024 * 1024);
```
I agree! And I usually complain about magic numbers... 🤦‍♂️ somehow I didn't notice it. Thanks, it's updated now.

LikeTheSalad commented on instrumentation/src/main/java/io/opentelemetry/android/config/PersistenceConfiguration.java:
```diff
@@ -0,0 +1,51 @@
+package io.opentelemetry.android.config;
+
+/** Configuration for disk buffering. */
+public final class PersistenceConfiguration {
```
Sounds good. The "persistence" term comes from the name they gave this feature in the Swift SDK and from there on it _persisted_ in my mind 😄 But yeah, without that context is just confusing to be calling it multiple names in here. I've updated it, thanks!

LikeTheSalad commented on instrumentation/src/main/java/io/opentelemetry/android/config/PersistenceConfiguration.java:
```diff
@@ -0,0 +1,51 @@
+package io.opentelemetry.android.config;
+
+/** Configuration for disk buffering. */
+public final class PersistenceConfiguration {
```
I don't really have a strong opinion on having all the params straight into the main config or grouping them within "subconfigs" if you will, though I'm a bit concerned about the intuitiveness/discoverability factor for someone who's not familiarized with the lib and wants to know all there is to configure about a specific feature. Now, this doesn't necessarily have to be the only way to address that issue, we could use a common prefix for all the config names related to a single feature, but that brings other concerns about having to enforce naming rules and also probably code placement rules (having all the related config methods together in one part of the file and so on) for expanding and maintaining the config code.

Now, the reason why it's not a strong opinion, at least not for now anyway, is because this is really only useful at a larger scale, in case there are multiple cases like this one, where several params are needed for a feature. And at the moment I can't think of another feature that would need this in the future (but this is a new project so I'm guessing that several features should come soon) and also we're talking about only 2 params for the disk buffering case (although we can technically expose more config options available in the [disk buffering lib](https://github.com/open-telemetry/opentelemetry-java-contrib/tree/main/disk-buffering#configuration), it's just that I tried to keep it simple for now).

Generally speaking, I try to make things as scalable as possible from the beginning, though I must admit that sometimes it backfires and I end up adding unnecessary things so, I like having these kinds of conversations because it helps making the line between _useful-for-scalability_ and _too-early/unnecessary-optimization_ a bit clearer.

LikeTheSalad commented on instrumentation/src/main/java/io/opentelemetry/android/internal/features/persistence/DiskManager.java:
```diff
@@ -0,0 +1,104 @@
+                config.getPersistenceConfiguration());
+    }
+
+    DiskManager(
```
Yeah, I just double-checked why is it like that and I remembered is because I needed it visible for tests (this code comes from the Elastic repo) so, depending on the test strategy we want to go with here, it might not be needed to have it package-private. Which brings me to the fact that I should update this PR to include tests since the overall approach was well received.

LikeTheSalad replied:
It's done now 👍 - I've added tests while also making the constructor private.

LikeTheSalad commented on instrumentation/src/main/java/io/opentelemetry/android/internal/features/persistence/DiskManager.java:
```diff
@@ -0,0 +1,104 @@
+        this.persistenceConfiguration = persistenceConfiguration;
+    }
+
+    public File getSignalsCacheDir() throws IOException {
```
Sure, it's updated now.

LikeTheSalad commented:
Thanks for the feedback! @breedx-splk @marandaneto 

I've addressed the comments and also added a couple of tests that were pending. I think it should be ready for another check.

breedx-splk reviewed (approved):
Good stuff!
